tools/ci-build/sdk-lockfiles/src/audit.rs

```diff
@@ -0,0 +1,430 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use crate::AuditArgs;
+use anyhow::bail;
+use anyhow::{Context, Result};
+use cargo_lock::dependency::graph::{Graph, NodeIndex};
+use cargo_lock::{package::Package, Lockfile};
+use petgraph::visit::EdgeRef;
+use smithy_rs_tool_common::git::find_git_repository_root;
+use smithy_rs_tool_common::here;
+use std::collections::{BTreeSet, HashSet};
+use std::env;
+use std::iter;
+use std::path::PathBuf;
+
+// A list of AWS runtime crate must be in sync with
+// https://github.com/smithy-lang/smithy-rs/blob/0f9b9aba386ea3063912a0464ba6a1fd7c596018/buildSrc/src/main/kotlin/CrateSet.kt#L42-L53
+// plus `aws-inlineable`
+const AWS_SDK_RUNTIMES: &[&str] = &[
+    "aws-config",
+    "aws-credential-types",
+    "aws-endpoint",
+    "aws-http",
+    "aws-hyper",
+    "aws-inlineable",
+    "aws-runtime",
+    "aws-runtime-api",
+    "aws-sig-auth",
+    "aws-sigv4",
+    "aws-types",
+];
+
+// A list of server runtime crates must be in sync with
+// https://github.com/smithy-lang/smithy-rs/blob/0f9b9aba386ea3063912a0464ba6a1fd7c596018/buildSrc/src/main/kotlin/CrateSet.kt#L85-L87
+const SERVER_SPECIFIC_RUNTIMES: &[&str] = &[
+    "aws-smithy-http-server",
+    "aws-smithy-http-server-python",
+    "aws-smithy-http-typescript",
+];
+
+fn new_dependency_for_aws_sdk(crate_name: &str) -> bool {
+    AWS_SDK_RUNTIMES.contains(&crate_name)
+        || crate_name == "inlineable"
```

Comment:
Depth first traversal for dependency check is performed in runtime lockfiles, importantly `rust-runtime/Cargo.lock` being one of them. The crate `inlineable` is [listed in that lockfile](https://github.com/smithy-lang/smithy-rs/blob/main/rust-runtime/Cargo.lock#L1998-L2024). This means that `inlineable`'s [dependencies](https://github.com/smithy-lang/smithy-rs/blob/main/rust-runtime/Cargo.lock#L2001-L2024) will be part of a set of crates used by a code-generated client SDK. That further means if an `inlineable`'s dependency were not listed in the SDK lockfile, compiling that client SDK crate would fetch the latest version of that unpinned crate from crates.io. We know that unpinned crate with its version subject to change is a problem for reproducible build.

This is the reason the `inlineable` crate also needs to be checked. There is room for improvement, though. A new dependency introduced in `inlineable` could only be used by the server SDK, in which case, that dependency is not needed by the client SDK, and the SDK lockfile doesn't need to include it. But this depth first traversal for dependency check is not smart enough to handle that case. We err on the side of caution by always considering a dependency introduced to the `inlineable` as required by the code-generated client SDK.

Reply:
Good question. Depth first traversal for dependency check is performed in runtime lockfiles, importantly `rust-runtime/Cargo.lock` being one of them. The crate `inlineable` is [listed in that lockfile](https://github.com/smithy-lang/smithy-rs/blob/main/rust-runtime/Cargo.lock#L1998-L2024). This means that `inlineable`'s [dependencies](https://github.com/smithy-lang/smithy-rs/blob/main/rust-runtime/Cargo.lock#L2001-L2024) will be part of a set of crates used by a code-generated client SDK, but _only when_ they are referred to by [CargoDependency.kt](https://github.com/smithy-lang/smithy-rs/blob/main/codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/rustlang/CargoDependency.kt).

In other words, I think your question just exposed a hole in the current audit mechanism in `sdk-lockfiles`. `sdk-lockfiles` (or maybe somewhere outside the tool, in which case we need to call that out as a tool's limitation) needs to check whether the SDK lockfile is behind `CargoDependency.kt` or not; if it's behind, audit should fail.

And we don't need to track `aws-inlineable` or `inlineable` here because knowing whether the SDK lockfile is behind `CargoDependency.kt` or not isn't possible as part of depth first traversal.

Reply:
Good question. Depth first traversal for dependency check is performed in runtime lockfiles, importantly `rust-runtime/Cargo.lock` being one of them. The crate `inlineable` is [listed in that lockfile](https://github.com/smithy-lang/smithy-rs/blob/main/rust-runtime/Cargo.lock#L1998-L2024). This means that `inlineable`'s [dependencies](https://github.com/smithy-lang/smithy-rs/blob/main/rust-runtime/Cargo.lock#L2001-L2024) will be part of a set of crates used by a code-generated client SDK, but _only when_ they are referred to by [CargoDependency.kt](https://github.com/smithy-lang/smithy-rs/blob/main/codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/rustlang/CargoDependency.kt).

We don't need to track `aws-inlineable` or `inlineable` here because knowing whether the SDK lockfile is behind `CargoDependency.kt` or not isn't possible as part of depth first traversal. That should probably be done by a parent script using this tool during the release process.